cpp/include/Ice/Communicator.h

```diff
@@ -340,23 +320,81 @@ namespace Ice
     private:
         static CommunicatorPtr create(InitializationData);
 
-        //
-        // Certain initialization tasks need to be completed after the
-        // constructor.
-        //
+        // Certain initialization tasks need to be completed after the constructor.
         void finishSetup(int&, const char*[]);
 
         [[nodiscard]] IceInternal::ReferencePtr _stringToProxy(std::string_view str) const;
         [[nodiscard]] IceInternal::ReferencePtr _propertyToProxy(std::string_view property) const;
 
+        /// @cond INTERNAL
         friend ICE_API CommunicatorPtr initialize(int&, const char*[], InitializationData);
-        friend ICE_API CommunicatorPtr initialize(StringSeq&, InitializationData);
         friend ICE_API CommunicatorPtr initialize(InitializationData);
+        /// @endcond
+
         friend ICE_API IceInternal::InstancePtr IceInternal::getInstance(const Ice::CommunicatorPtr&);
         friend ICE_API IceInternal::TimerPtr IceInternal::getInstanceTimer(const Ice::CommunicatorPtr&);
 
         const IceInternal::InstancePtr _instance;
     };
+
+    /// A helper class that uses Resource Acquisition Is Initialization (RAII) to hold a communicator instance, and
+    /// automatically destroy this instance when the holder goes out of scope.
+    /// @headerfile Ice/Ice.h
+    class ICE_API CommunicatorHolder
+    {
+    public:
+        /// Default constructor.
+        CommunicatorHolder() = default;
+
+        /// Constructs a CommunicatorHolder that adopts an existing communicator.
+        /// @param communicator The communicator to adopt.
+        CommunicatorHolder(CommunicatorPtr communicator) noexcept;
+
+        /// Constructs a CommunicatorHolder for a new communicator created using an `Ice::initialize` overload.
+        /// @tparam T The types of the arguments to pass to the `Ice::initialize` function.
+        /// @param args The arguments to pass to the `Ice::initialize` function.
+        template<class... T>
+        explicit CommunicatorHolder(T&&... args) : _communicator(std::move(initialize(std::forward<T>(args)...)))
+        {
+        }
+
+        /// Move constructor. Constructs a CommunicatorHolder with the contents of @p other using move semantics.
+        /// @param other The holder to move from.
+        CommunicatorHolder(CommunicatorHolder&& other) = default;
+
+        CommunicatorHolder(const CommunicatorHolder&) = delete;
+
+        /// Assignment operator. Destroys the current communicator (if any) and adopts a new communicator.
+        /// @param communicator The communicator to adopt.
+        CommunicatorHolder& operator=(CommunicatorPtr communicator) noexcept;
+
+        CommunicatorHolder& operator=(const CommunicatorHolder&) noexcept = delete;
```

Comment:
Do we need tp document the const variant?

Reply:
Which const variant?

In general, we don't document `=delete`  functions.

Reply:
missed it was deleted.